File: Python/I/11-ZOO/5/dragdroplineedit.py
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *
from dragdroplabel import *

class DragDropLineEdit(QLineEdit):
    tag = -1

    def dragEnterEvent(self, event):
        event.accept()
        self.setStyleSheet('background-color: yellow;')

    def dropEvent(self, event):
        src = event.source()
        if src.tag >= self.tag:
            self.setText(src.text())
            self.tag = src.tag
        src.setVisible(False)
        self.setStyleSheet('background-color: #f0f0f0;')

    def dragLeaveEvent(self, event):
        self.setStyleSheet('background-color: #f0f0f0;')

File: C++/I/11-ZOO/5/dragdroplineedit.cpp
#include "dragdroplineedit.h"
#include "dragdroplabel.h"

DragDropLineEdit::DragDropLineEdit(QWidget *parent):
    QLineEdit(parent){}

void DragDropLineEdit::dragEnterEvent(QDragEnterEvent *event)
{
    event->accept();
    setStyleSheet("background-color: yellow;");
}

void DragDropLineEdit::dropEvent(QDropEvent *event)
{
    DragDropLabel *lb = (DragDropLabel*)event->source();
    if (lb->tag() >= tag())
    {
        setText(lb->text());
        setTag(lb->tag());
    }
    lb->setVisible(false);
    setStyleSheet(styleSheet().append("background-color: #f0f0f0;"));
}

void DragDropLineEdit::dragLeaveEvent(QDragLeaveEvent*)
{
    setStyleSheet("background-color: #f0f0f0;");
}
